nav2_map_server/include/nav2_map_server/map_saver.hpp

```diff
@@ -48,24 +78,54 @@ class MapSaver : public rclcpp::Node
    * @brief Returns the promise as shared future
    * @return a shared future copy of save_next_map_promise
    */
-  std::shared_future<void> map_saved_future() {return save_next_map_promise.get_future().share();}
+  std::shared_future<void> map_saved_future() {
+    return save_next_map_promise.get_future().share();
+  }
+
+  void new_promise() {
+    save_next_map_promise = std::promise<void>();
+  }
+
+  /**
+   * @brief Writes map data to file. Static API function.
+   * @param map Occupancy grid data
+   * @param save_parameters Map saving parameters
+   * @return true or false
+   */
+  static bool saveMapToFile(
+    const nav_msgs::msg::OccupancyGrid & map,
+    SaveParameters & save_parameters);
 
 protected:
-  rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::ConstSharedPtr map_sub_;
+  /**
+   * @brief Checks map saving parameters for consistency
+   * @param save_parameters Map saving parameters.
+   * NOTE: map saving paramters could be updated during function
+   * execution.
+   * @throw std::exception in case of inconsistent parameters
+   */
+  static void checkSaveParameters(SaveParameters & save_parameters);
 
   /**
-   * @brief Writes map data to file
+   * @brief Tries to write map data into a file
    * @param map Occupancy grid data
+   * @param save_parameters Map saving parameters
+   * @throw std::expection in case of problem
    */
-  void try_write_map_to_file(const nav_msgs::msg::OccupancyGrid & map);
+  static void tryWriteMapToFile(const nav_msgs::msg::OccupancyGrid & map,
+    const SaveParameters & save_parameters);
 
+  // Promise for class spinning synchronization
   std::promise<void> save_next_map_promise;
 
-  std::string image_format;
-  std::string mapname_;
-  int threshold_occupied_;
-  int threshold_free_;
-  nav2_map_server::MapMode map_mode;
+  // Topic that MapSaver will listen to save the map
+  std::string map_topic_;
+
+  // Map topic subscriber
+  rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::ConstSharedPtr map_sub_;
```

Comment:
This doesn't need to be a class variable, it can be created and destroyed in the save map service callback so you don't need to retain state

Reply:
Fixed: https://github.com/ros-planning/navigation2/commit/84cff7d8cf43443e23bc4852129791d665f6e5c7#diff-99885e6562ca50edd8cc80db6362d317R162